<a href="https://colab.research.google.com/github/yogeshrana7/BDDA2/blob/main/credit_card_fraud_detection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [63]:
import numpy as np
import pandas as pd
import seaborn as sns
from sklearn.metrics import accuracy_score, classification_report
from  sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

In [64]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [65]:
file_transaction = pd.read_csv("/content/drive/MyDrive/creditcard.csv")
file_transaction.head(5)

Time        V1        V2        V3  ...       V27       V28  Amount  Class
0   0.0 -1.359807 -0.072781  2.536347  ...  0.133558 -0.021053  149.62      0
1   0.0  1.191857  0.266151  0.166480  ... -0.008983  0.014724    2.69      0
2   1.0 -1.358354 -1.340163  1.773209  ... -0.055353 -0.059752  378.66      0
3   1.0 -0.966272 -0.185226  1.792993  ...  0.062723  0.061458  123.50      0
4   2.0 -1.158233  0.877737  1.548718  ...  0.219422  0.215153   69.99      0

[5 rows x 31 columns]

In [66]:
file_transaction.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [67]:
file_transaction.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [68]:
file_transaction['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

In [69]:
file_legit_transaction = file_transaction[file_transaction.Class == 0] 
file_fraud_transaction = file_transaction[file_transaction.Class == 1]
#print(file_fraud_transaction.to_string()) # print the new sub table

In [70]:
print(file_legit_transaction.shape)
print(file_fraud_transaction.shape)

(284315, 31)
(492, 31)


In [71]:
file_legit_transaction.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [72]:
file_fraud_transaction.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [73]:
file_transaction.groupby('Class').mean()

Time        V1        V2  ...       V27       V28      Amount
Class                                    ...                                
0      94838.202258  0.008258 -0.006271  ... -0.000295 -0.000131   88.291022
1      80746.806911 -4.771948  3.623778  ...  0.170575  0.075667  122.211321

[2 rows x 30 columns]

In [74]:
file_legit_transaction_sample = file_legit_transaction.sample(n=492)

In [75]:
new_file_transaction= pd.concat([file_legit_transaction_sample, file_fraud_transaction], axis = 0)

In [76]:
new_file_transaction.shape

(984, 31)

In [77]:
new_file_transaction.head(4)

Time        V1        V2  ...       V28  Amount  Class
271776  164741.0 -0.923377  1.752366  ...  0.235718    1.79      0
171064  120483.0 -0.632781 -1.511681  ...  0.152654   10.30      0
283728  171829.0  2.070626  0.176873  ... -0.029157    9.99      0
224172  143681.0  2.228097 -0.609542  ... -0.096551   40.00      0

[4 rows x 31 columns]

In [78]:
new_file_transaction['Class'].value_counts()

1    492
0    492
Name: Class, dtype: int64

In [79]:
new_file_transaction.groupby('Class').mean()

Time        V1        V2  ...       V27       V28      Amount
Class                                    ...                                
0      91719.136179 -0.035207 -0.103643  ...  0.014390 -0.021507   92.710346
1      80746.806911 -4.771948  3.623778  ...  0.170575  0.075667  122.211321

[2 rows x 30 columns]

In [80]:
X = new_file_transaction.drop(columns = 'Class', axis=1)
Y = new_file_transaction['Class']

In [81]:
X.shape

(984, 30)

In [82]:
Y.shape

(984,)

In [89]:
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2)


In [90]:
model = LogisticRegression()

In [91]:
model.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [92]:
train_predictions = model.predict(X_train)
train_accuracy = accuracy_score(train_predictions, y_train)
print(train_accuracy)

0.9529860228716646


In [93]:
test_predictions = model.predict(X_test)
test_accuracy = accuracy_score(test_predictions, y_test)
print(test_accuracy)

0.9086294416243654


In [94]:
print(classification_report(y_test,test_predictions))

              precision    recall  f1-score   support

           0       0.88      0.95      0.91        99
           1       0.94      0.87      0.90        98

    accuracy                           0.91       197
   macro avg       0.91      0.91      0.91       197
weighted avg       0.91      0.91      0.91       197



#####  Conclusion : Model have achieved an accuracy of approx 91% .